AdminServer/appscale/admin/__init__.py

```diff
@@ -130,20 +132,212 @@ def wait_for_delete(operation_id, http_port):
-
-    if not utils.port_is_open(options.login_ip, http_port):
+    to_remove = []
+    for http_port in ports:
+      # If the port is open, continue to process other ports.
+      if utils.port_is_open(options.login_ip, int(http_port)):
+        continue
+      # Otherwise one more port has finished and remove it from the list of
+      # ports to check.
+      finished += 1
+      to_remove.append(http_port)
+    ports = [p for p in ports if p not in to_remove]
+    if finished == len(ports_to_close):
       break
 
     yield gen.sleep(1)
 
   operation.finish()
 
 
+class BaseVersionHandler(BaseHandler):
+
+  def get_version(self, project_id, service_id, version_id):
+    """ Fetches a version node.
+
+    Args:
+      project_id: A string specifying a project ID.
+      service_id: A string specifying a service ID.
+      version_id: A string specifying a version ID.
+    Returns:
+      A dictionary containing version details.
+    """
+    version_node = constants.VERSION_NODE_TEMPLATE.format(
+      project_id=project_id, service_id=service_id, version_id=version_id)
+
+    try:
+      version_json, _ = self.zk_client.get(version_node)
+    except NoNodeError:
+      raise CustomHTTPError(HTTPCodes.NOT_FOUND, message='Version not found')
+
+    return json.loads(version_json)
+
+  @gen.coroutine
+  def start_delete_version(self, project_id, service_id, version_id):
+    """ Starts the process of deleting a version by calling stop_version on
+    the AppController and deleting the version node. Returns the version's
+    port that will be closing, the caller should wait for this port to close.
+
+    Args:
+      project_id: A string specifying a project ID.
+      service_id: A string specifying a service ID.
+      version_id: A string specifying a version ID.
+    Returns:
+      The version's port.
+    """
+    version = self.get_version(project_id, service_id, version_id)
+    try:
+      http_port = int(version['appscaleExtensions']['httpPort'])
+    except KeyError:
+      raise CustomHTTPError(HTTPCodes.NOT_FOUND,
+                            message='Version serving port not found')
+
+    version_node = constants.VERSION_NODE_TEMPLATE.format(
+      project_id=project_id, service_id=service_id, version_id=version_id)
+
+    yield self.thread_pool.submit(self.version_update_lock.acquire)
+    try:
+      try:
+        self.zk_client.delete(version_node)
+      except NoNodeError:
+        pass
+    finally:
+      self.version_update_lock.release()
+
+    version_key = VERSION_PATH_SEPARATOR.join([project_id, service_id,
+                                               version_id])
+    try:
+      self.acc.stop_version(version_key)
+    except AppControllerException as error:
+      message = 'Error while stopping version: {}'.format(error)
+      raise CustomHTTPError(HTTPCodes.INTERNAL_ERROR, message=message)
+
+    raise gen.Return(http_port)
+
+
+class ProjectHandler(BaseVersionHandler):
```

Comment:
ProjectHandler doesn't look like a particular case of version handler. But instead it just uses some common functionality.
I'd suggest to avoid using subclassing here and move start_delete_version outside.
Your implementation looks good and seems to correspond to popular practice of reusing code. But I want to leave this remark here and hear what you think about it=)

Reply:
My reason for having them extend the BaseVersionHandler is because at the root of it the ProjectHandler and ServiceHandler undeploy versions, they just loop through and undeploy all the versions of the project or the service.